### ISSUE TITLE
TestTableCreateSnapshotAt fails in development build

### ISSUE THREAD

**Issue by lukepalmer** [VictoriaMetrics/VictoriaMetrics#4272]
### Describe the bug

When running "make test" I consistently get a test failure for mergeset.

The automated docker builds of course pass, which means there is something about my environment at work here.  This is still important, because I don't (can't) run victoriametrics in production with docker.

I'm happy to help dig but would love a starting point from someone who knows things about this test.

### To Reproduce

'make test'

### Version

HEAD, both single and cluster

$ go version
go version go1.20.4 linux/amd64
$ cat /etc/redhat-release
Rocky Linux release 8.7 (Green Obsidian)

### Logs

--- FAIL: TestTableCreateSnapshotAt (0.14s)
    table_test.go:153: unexpected item found for key="item 14" in snapshot1; got "item 140060"

**Comment by hagen1778** [comment]
Hi @lukepalmer!

Could you please verify before running the test that path `lib/mergeset/TestTableCreateSnapshotAt` doesn't exist? Could it be that it contains some old files?

**Comment by lukepalmer** [comment]
Thanks for looking @hagen1778!
Verified. This reproduces for me 100% of the time with a completely fresh git clone.

If you'd have some pointers I'm happy to dig more.

**Comment by hagen1778** [comment]
Looks like there is a bug in a testing function named `DebugFlush`. It is responsible to persist data from in-memory to the disk for testing purposes. Apparently, it doesn't wait for all data to persist which causes this problem.
Would adding artificial delay `time.Sleep(time.Seond*5)` after this function call in that test solve your problem?

**Comment by lukepalmer** [comment]
Great- That does make the test pass.

Interestingly enough, this could also explain some non-determinism in an external integration test I have that uses the  /internal/force_flush endpoint.

**Comment by dmitryk-dk** [comment]
Hi @lukepalmer ! I have created in gcloud my test instance, which is :
<img width="644" alt="Screenshot 2023-05-12 at 15 08 14" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/12189570/db266fdd-bce2-44f5-bd2c-116f3abacfa1">
And run all tests from the root of the VictoriaMetrics folder via `make test`. And in my installation, all tests are passed successfully.
Is there a difference between installations?
What I have
```
Machine type e2-standard-4
CPU platform Intel Broadwell
Architecture x86/64
```

```
Used disk:

dk | rocky-linux-8-optimized-gcp-v20230509 | SCSI | 40 | dk-dev | Standard persis
```
The image from the my run
<img width="943" alt="Screenshot 2023-05-12 at 15 11 56" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/12189570/494217b2-cabe-4f1c-9ccc-dd7256e8315d">

**Comment by dmitryk-dk** [comment]
> @dmitryk-dk it doesn't mean that function works as expected. Try to increase number of elements in the test case by 10-20x and re-run it again. Once the write creates more than 1 part - test starts to fail.

Yes, you are right! Need some more time to investigate

**Comment by lukepalmer** [comment]
I found something intresting: I think this is related to parallelism. My development machine is a virtualized 64 cores. If I limit the cores available like this:

taskset -c 12-20 make test

The problem doesn't reproduce.

**Comment by valyala** [comment]
> Looks like there is a bug in a testing function named DebugFlush. It is responsible to persist data from in-memory to the disk for testing purposes. Apparently, it doesn't wait for all data to persist which causes this problem.

The `DebugFlush` function doesn't persist data to disk. It just makes sure that the recently ingested data is available to search. The commit 09b403d38a265af362fabc269b2345e63bc42af6 makes `DebugFlush()` comments less misleading.

The `TestTableCreateSnapshotAt()` test may be flaky because of the following reason.

VictoriaMetrics stores recently ingested data into in-memory buffers. The data in these buffers isn't visible to search.
The in-memory buffers are converted to searchable in-memory parts either every second or when they become full, whichever comes first. After that the data becomes visible for search.
The in-memory parts are converted into file-based parts with interval, which can be specified via `-inmemoryDataFlushInterval` command-line flag. By default it is set to 5 seconds. See [these docs](https://docs.victoriametrics.com/#storage) for details.

VictoriaMetrics creates [snapshots](https://docs.victoriametrics.com/#how-to-work-with-snapshots) only from file-based parts, e.g. it ignores in-memory parts. This works in most practical cases when a constant stream of new data is ingested into VictoriaMetrics and snapshots are created on a regular basis (e.g. once per hour / day / week) in order to be able to recover from these snapshots at the particular point in time. There is no big deal if a few seconds of recently ingested data isn't available in the snapshot in this practical case.

VictoriaMetrics doesn't store in-memory parts into snapshots because it may be resource-consuming and time-consuming operation. E.g. it needs an additional CPU and disk IO resources for converting of in-memory parts to file-based parts. It may be also tricky to make sure that all the in-memory parts are flushed to disk during snapshot creating, since some in-memory parts can hang for indefinite duration because of [background merges](https://docs.victoriametrics.com/#storage), while new in-memory parts may appear during this time because of constant stream of newly ingested data.

There are valid use case when it is required that snapshots contain all the recently ingested data - these use cases are named `tests`, like the one at `TestTableCreateSnapshotAt()`. And it should be fixed, while keeping in mind the most practical use case described above.

**Comment by valyala** [comment]
Historical note - VictoriaMetrics prior v1.90.0 was trying to flush all the in-memory parts to disk before creating a snapshot. This wasn't working well under constant data ingestion - [see the corresponding issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3551). The commit 6460475e3bb488086046cb5fdce3915206d731ef fixed the issue in v1.90.0 . Unfortunately it also led to flaky `TestTableCreateSnapshotAt()` test.

**Comment by valyala** [comment]
FYI, the first suggestion on how to fix the `TestTableCreateSnapshotAt()` from the https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4272#issuecomment-1550221840 has been implemented in the commit 3cbc0975f6d9fb197b825ac2406fb29129c10c52 .

@lukepalmer , could you verify whether the test now succeeds all the time in your environments?

**Comment by lukepalmer** [comment]
Looks great now, thanks all.
